contracts/puppeteer-initia/src/contract.rs

```diff
@@ -0,0 +1,1331 @@
+            type_url: "/initia.mstaking.v1.MsgUndelegate".to_string(),
+            value: drop_proto::proto::initia::mstaking::v1::MsgUndelegate {
+                delegator_address: delegator.to_string(),
+                validator_address: validator.to_string(),
+                amount: vec![drop_proto::proto::cosmos::base::v1beta1::Coin {
+                    denom: config.remote_denom.to_string(),
+                    amount: amount.to_string(),
+                }],
+            }
+            .to_bytes()?,
+        })
+    }
+
+    let grant_msg = MsgExec {
+        grantee: delegator,
+        msgs: undelegation_msgs,
+    };
+
+    let any_msgs: Vec<neutron_sdk::bindings::types::ProtobufAny> =
+        vec![prepare_any_msg(grant_msg, "/cosmos.authz.v1beta1.MsgExec")?];
+
+    let submsg = compose_submsg(
+        deps.branch(),
+        config.clone(),
+        any_msgs,
+        Transaction::Undelegate {
+            interchain_account_id: ICA_ID.to_string(),
+            denom: config.remote_denom,
+            batch_id,
+            items,
+        },
+        reply_to,
+        ReplyMsg::SudoPayload.to_reply_id(),
+    )?;
+
+    Ok(Response::default().add_submessages(vec![submsg]))
+}
+
+fn execute_redelegate(
+    mut deps: DepsMut<NeutronQuery>,
+    info: MessageInfo,
+    validator_from: String,
+    validator_to: String,
+    amount: Uint128,
+    reply_to: String,
+) -> ContractResult<Response<NeutronMsg>> {
+    let puppeteer_base = Puppeteer::default();
+    deps.api.addr_validate(&reply_to)?;
+    let config: Config = puppeteer_base.config.load(deps.storage)?;
+    validate_sender(&config, &info.sender)?;
+    puppeteer_base.validate_tx_idle_state(deps.as_ref())?;
+    let delegator = puppeteer_base.ica.get_address(deps.storage)?;
+    let redelegate_msg = MsgBeginRedelegate {
+        delegator_address: delegator,
+        validator_src_address: validator_from.to_string(),
+        validator_dst_address: validator_to.to_string(),
+        amount: vec![ProtoCoin {
+            denom: config.remote_denom.to_string(),
+            amount: amount.to_string(),
+        }],
+    };
+
+    let submsg = compose_submsg(
+        deps.branch(),
+        config.clone(),
+        vec![prepare_any_msg(
+            redelegate_msg,
+            "/initia.mstaking.v1.MsgBeginRedelegate",
+        )?],
+        Transaction::Redelegate {
+            interchain_account_id: ICA_ID.to_string(),
+            validator_from,
+            validator_to,
+            denom: config.remote_denom,
+            amount: amount.into(),
+        },
+        reply_to,
+        ReplyMsg::SudoPayload.to_reply_id(),
+    )?;
+
+    Ok(Response::default().add_submessages(vec![submsg]))
+}
+
+fn execute_tokenize_share(
+    _deps: DepsMut<NeutronQuery>,
+    _info: MessageInfo,
+    _validator: String,
+    _amount: Uint128,
+    _reply_to: String,
+) -> ContractResult<Response<NeutronMsg>> {
+    panic!("Not implemented")
```

Comment:
unimplemented!() is better
```suggestion
    unimplemented!()
```